task splitting: change additive accumulation to multiplicative

StefanKarpinski opened the pull request:
Issue raised [here](https://discourse.julialang.org/t/linear-relationship-between-xoshiro-tasks/110454). Given this definition, look at output values:
```jl
using .Threads

macro fetch(ex) :(fetch(@spawn($(esc(ex))))) end

function taskCorrelatedXoshiros(+ = +)
    r11, r10 = @fetch (@fetch(rand(UInt64)), rand(UInt64))
    r01, r00 = (@fetch(rand(UInt64)), rand(UInt64))
    r01 + r10 - (r00 + r11)
end
```
Before:
```jl
julia> sort!(unique(taskCorrelatedXoshiros() for _ = 1:1000))
9-element Vector{UInt64}:
 0x0000000000000000
 0x0000000000000001
 0x00000000007fffff
 0x0000000000800000
 0x0000000000800001
 0xffffffffff7fffff
 0xffffffffff800000
 0xffffffffff800001
 0xffffffffffffffff
```
After:
```jl
julia> sort!(unique(taskCorrelatedXoshiros() for _ = 1:1000))
1000-element Vector{UInt64}:
 0x000420647a085960
 0x0038c5b889b585c6
 0x007b29fae8107ff7
 0x00e73b9e883ac1c8
                  ⋮
 0xfe68be9c0dde1e88
 0xfeca042354218c35
 0xfeeb8203e470c96b
 0xfff5dbb8771307b9

julia> sort!(unique(taskCorrelatedXoshiros(*) for _ = 1:1000))
1000-element Vector{UInt64}:
 0x00126f951c1e56dc
 0x0025a82477ffac08
 0x002dd82c9986457a
 0x00a713c4d56a3dbc
                  ⋮
 0xfe2e40a5b345095e
 0xfe77b90881967436
 0xfea2559be63f1701
 0xff88b5b28cefac5f
```

vtjnash reviewed (approved):
Where do these new `a[1:4]` values come from?

Seelengrab commented:
I guess this part of the explanation/doc above `jl_rng_split` should be adjusted too:

> It also means that when accumulating the dot product incrementally, as
described in SplitMix, we don't need to multiply weights by anything, we simply
add the random weight for the current task tree depth to the parent's dot
product to derive the child's dot product.

StefanKarpinski commented:
> Where do these new `a[1:4]` values come from?

As the comment says, they are "random additive offsets".

StefanKarpinski commented:
> I guess this part of the explanation/doc above `jl_rng_split` should be adjusted too:

Yes, I'm not adjusting and comment text until we settle on something.

oscardssmith commented:
Does this change notably affect the task spawn time? (I'm assuming not, because add, multiply and byteswap are all really fast, but good to make sure). Otherwise, this looks like a good change.

StefanKarpinski commented:
I get identical timings between master and this rebased onto master—about 65ns. Would be good if someone could benchmark on different systems. I'm on an M2 system, so x86 would be helpful.

carstenbauer commented:
Just ran a quick and dirty benchmark on one of our compute nodes (AMD Zen 3).

```julia
# master (a1db8daafd*)
julia> @btime @spawn(nothing);
  65.741 ns (4 allocations: 480 bytes)

# this PR rebased onto master
julia> @btime @spawn(nothing);
  66.098 ns (4 allocations: 480 bytes)
```

StefanKarpinski committed:
jl_rng_split: fix linear possible correlation between four tasks

In response to the demonstrated ease of constructing linearly correlated
task states, and thus linearly related task RNG outputs, this change
departs from using linear mixing for the compression function, which is
simply dot product in those RNGs. I've shown that we can use any mixing
function that's doubly bijective. This change uses `(2c+1)(2w+1)>>1` to
mix the LCG state and the xoshiro register state and _then_ applies the
PCG output function, which is designed to mask linearity.

StefanKarpinski commented on src/task.c:
```diff
@@ -856,202 +856,143 @@ guaranteed to avoid collisions between the RNG streams of all tasks. The main
+pseudorandom weights are multiplied by corresponding task coordinate and added
+up. While this does provably make collisions as unlikely as randomly assigned
+task seeds, this linear construction can be used to create linearly correlated
+states between states. However, it turns out that the compression construction
```
```suggestion
states between tasks. However, it turns out that the compression construction
```

StefanKarpinski committed:
Update src/task.c

StefanKarpinski commented on src/task.c:
```diff
@@ -856,202 +856,143 @@ guaranteed to avoid collisions between the RNG streams of all tasks. The main
+PCG-RXS-M-XS-64 RNG and allows it to pass statistical RNG test suites despite
+having the same size state and output. In particular, since this mixing function
+is highly non-associative and non-linear, we (hopefully) don't have any
+discernable relationship between these values:
```
```suggestion
discernible relationship between these values:
```

StefanKarpinski commented on src/task.c:
```diff
@@ -856,202 +856,143 @@ guaranteed to avoid collisions between the RNG streams of all tasks. The main
+undesirable and highlights a hazard of the plain dot product construction.
+
+As in DotMix and SplitMix, each task is assigned unique task "pedigree"
+coordiates. Our pedigree construction is a bit different and uses only binary
```
```suggestion
coordinates. Our pedigree construction is a bit different and uses only binary
```

StefanKarpinski committed:
Apply suggestions from code review

StefanKarpinski commented:
Ok, this is ready to go from my perspective.

KristofferC commented:
I want to see what package's tests "broke" from the RNG stream changing here.

@nanosoldier `runtests(; vs = "@c80a9647166d59782eff6be8c75a5052e17483d7")`
